File: src/main/java/me/centralhardware/znatoki/telegram/statistic/eav/Property.java
package me.centralhardware.znatoki.telegram.statistic.eav;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import me.centralhardware.znatoki.telegram.statistic.eav.jackson.TypeDeserializer;
import me.centralhardware.znatoki.telegram.statistic.eav.jackson.TypeSerializer;
import me.centralhardware.znatoki.telegram.statistic.eav.types.Type;

public record Property(
        String name,
        @JsonSerialize(using = TypeSerializer.class)
        @JsonDeserialize(using = TypeDeserializer.class)
        Type type,
        String value) {

    public Property(String name, Type type){
        this(name,type,"");
    }

    public Property withValue(String value){
        return new Property(name(), type(), value );
    }


}

File: src/main/kotlin/me/centralhardware/znatoki/telegram/statistic/eav/Property.kt
package me.centralhardware.znatoki.telegram.statistic.eav

import kotlinx.serialization.Serializable
import me.centralhardware.znatoki.telegram.statistic.eav.json.TypeAdapter
import me.centralhardware.znatoki.telegram.statistic.eav.types.Type

@Serializable
data class Property(
    val name: String,
    @Serializable(with = TypeAdapter::class) val type: Type,
    var value: String? = "",
) {
    fun withValue(value: String) = this.apply { this.value = value }
}
